### ISSUE TITLE
Leaflet 1.7.1 causes 2 click events to be emitted by Leaflet Controls in Safari

### ISSUE THREAD

**Issue by dankarran** [Leaflet/Leaflet#7255]
**Steps to reproduce**
- set up a basic map, with a control that has a button in it, which listens for click events and posts the events to the console
- click the button in the control using Safari on a Mac
- see two click events in the console, one with `isTrusted:true` and another with `isTrusted:false` attributes

(Using 'force touch' to click the button on a trackpad results in only a single click event being emitted, but it also triggers the Safari preview popup.)

**Expected behavior**

There should only be one click event emitted, with the `isTrusted:true` attribute.

Would Leaflet be able to safely filter out the click events with `isTrusted:false`?

**Current behavior**

In version 1.6 the expected behaviour is observed. In version 1.7.1 Chrome exhibits the expected behaviour, but Safari doesn't.

The differences between the two events are included below.

MouseEvent 1:
```
MouseEvent

_simulated: true
composed: false
isTrusted: false
layerX: 15
layerY: 28
offsetX: 15
offsetY: 28
pageX: 410
pageY: 46
screenX: 1390
screenY: 189
timeStamp: 499337
webkitForce: 0
x: 410
y: 38
```

MouseEvent 2:
```
MouseEvent

composed: true
isTrusted: true
layerX: 15
layerY: 20
offsetX: 15
offsetY: 20
pageX: 410
pageY: 38
screenX: 1390
screenY: 189
timeStamp: 538265
webkitForce: 1
x: 410
y: 38
```

`_simulated`, `composed`, `isTrusted` and perhaps `webkitForce` seem to be the key differences, but it also seems odd that there are different timestamps and Y dimensions between the two events.

**Environment**
- Leaflet version: 1.7.1
- Browser (with version): Safari 13.1.2 
- OS/Platform (with version): OS X 10.13.6 and 10.15.6

**Additional context**

I noticed this after OpenStreetMap updated to the latest version of Leaflet, where the issue manifests itself by preventing some of the map tools to be used as the first event toggles a tool on and the second toggles it off. Related issue posted at https://github.com/openstreetmap/openstreetmap-website/issues/2811

**Minimal example reproducing the issue**

Please see https://codepen.io/dankarran/pen/JjXMXzd

**Comment by IvanSanchez** [comment]
> Minimal example reproducing the issue

For the sake of clarity: that example is using

```
    L.DomEvent.on(container, "click", function(e) {
```

**Comment by IvanSanchez** [comment]
> Would Leaflet be able to safely filter out the click events with `isTrusted:false`?

It would be possible to do so, in `addOne()` in `DomEvent.js`, around https://github.com/Leaflet/Leaflet/blob/4f32a5e83525a3a42980c974e48712b058510eb5/src/dom/DomEvent.js#L109

With a new branch in that if-then-else chain, something probably like

```js
} else if (Browser.safari && !Browser.mobile && type === 'click') {
  handler = function (ev) {
    if (ev.isTrusted) { originalHandler(ev); }
  }
  obj.addEventListener('click', handler, false);
} // etc
```

However, I do not own any Apple hardware (at the moment), so I'm unable to test this; and I don't know if this might impact previous versions of desktop Safari.

**Comment by IvanSanchez** [comment]
The simulated event comes from https://github.com/Leaflet/Leaflet/blob/4f32a5e83525a3a42980c974e48712b058510eb5/src/map/handler/Map.Tap.js#L100

So I suggest changing the condition at https://github.com/Leaflet/Leaflet/blob/4f32a5e83525a3a42980c974e48712b058510eb5/src/map/handler/Map.Tap.js#L87

into something like 

```js
if (this._fireClick && e && e.changedTouches && !(Browser.safari && !Browser.mobile)) { 
```

However, I do not own any Apple hardware (at the moment), so I'm unable to test this; and I don't know if this might impact previous versions of desktop Safari, or other browsers (Does the bug maybe reproduce on firefox mobile?).

**Comment by IvanSanchez** [comment]
I think this bug has been introduced at

https://github.com/Leaflet/Leaflet/pull/7010/files#diff-c2e3d5ef35bf1c4a7c6b549ddf734725R134

@johnd0e , can you have a look? Maybe you introduced that change to account for iOS but unintentionally affected desktop macOS?

**Comment by dankarran** [comment]
Thanks for looking into it @IvanSanchez. I'm happy to help testing any proposed changes on Safari on Mac or iOS.

Just to add, there is a [related problem](https://github.com/openstreetmap/openstreetmap-website/issues/2811#issuecomment-687619650) in Safari on iOS, but it seems to behave slightly differently. I haven't done much testing on that.

**Comment by johnd0e** [comment]
> @johnd0e , can you have a look? Maybe you introduced that change to account for iOS but unintentionally affected desktop macOS?

That was definitely meant to address iOS issues, see discussion: https://github.com/Leaflet/Leaflet/pull/7010.
And we use `safari` property intentionally, in order not to be tooo specific.

Possible quickfixes:
- introduce `Browser.ios` property
- or use `Browser.safari && Browser.mobile`

But what if tomorrow we will get macbook with touchscren? That will break it all again.

**Comment by johnd0e** [comment]
So I propose to apply quickfix (`Browser.safari && Browser.mobile`) ASAP, and then discuss ultimate solution here: #6980

**Comment by IvanSanchez** [comment]
>     * or use `Browser.safari && Browser.mobile`

I like this approach (as a "quick & dirty" fix).

> But what if tomorrow we will get macbook with touchscren? That will break it all again.

You know that my answer, in the end, is gonna be #7200 :-)

**Comment by johnd0e** [comment]
> You know that my answer, in the end, is gonna be #7200 :-)

I am not so sure)
But anyway, I'd prefer to fix all we can staying in current approach, and then move further (to #7200 or whatever).

**Comment by systemed** [comment]
Unsure whether this is the same issue (so please forgive tagging on existing ticket), but shift-drag to zoom in is also broken in 1.7.1 on desktop Safari.

**Comment by IvanSanchez** [comment]
@systemed Can you please try shift-dragging using https://s3.amazonaws.com/leafletjs-cdn/content/leaflet/master/leaflet-src.js ? (e.g. https://plnkr.co/edit/v4GzrCYSxpIO1aOo ) That's a build from `master` that already includes the changes from #7260.

**Comment by systemed** [comment]
Spot on - that works fine. 👍

**Comment by johnd0e** [comment]
As appeared, the same issue affects not only desktop Safari, but also legacy Edge (and who knows what else).
AFAIK atm `tap` handler is needed for iOS **only**.

So... May be we should change `map.options.tap` default to `false`.

**Comment by dankarran** [comment]
I've updated my [codepen](https://codepen.io/dankarran/pen/JjXMXzd) to use the code built from master, and also to display the events to make it easier to test on mobile.

It's working well on desktop Safari, but unfortunately, it looks like mobile Safari is still firing off two click events.

**Comment by dankarran** [comment]
If the JS at https://s3.amazonaws.com/leafletjs-cdn/content/leaflet/master/leaflet-src.js is the latest version, it's still firing two click events on iOS Safari.

**Comment by johnd0e** [comment]
@dagjomar 
Are you able to inspect those events?

It may be different issue.

Because on `touchstart` (and `pointerdown`) we prevent `click` here:
https://github.com/Leaflet/Leaflet/blob/dda26ba96d7dc79213818d1bcc0565b6ac3c69c4/src/map/handler/Map.Tap.js#L43
And I do not see any reason why that might not work.

**Comment by dankarran** [comment]
@johnd0e I'll see if I can get some more details out, and let you know. Would it make a difference what type of element was used (e.g. `a` or `button` which would have their own click handlers vs `div` which wouldn't)?

**Comment by dankarran** [comment]
I've updated the codepen to log some simplified event details to the console on there. The events on iOS seem to be the same as seen on the desktop before this update:

```
type: click
isTrusted: false
composed: false
_simulated: true
webkitForce: 0
```

```
type: click
isTrusted: true
composed: true
webkitForce: 1
```

Also, `a` and `div` elements behave the same, and adding `e.preventDefault()` in my event handler doesn't solve it.

Let me know if there's anything else you'd like me to try @johnd0e?

**Comment by dankarran** [comment]
Is it possible that the following lines aren't needed, if Safari is firing its own click events on mouseup?

https://github.com/Leaflet/Leaflet/blob/4f32a5e83525a3a42980c974e48712b058510eb5/src/map/handler/Map.Tap.js#L98-L101

**Comment by dankarran** [comment]
By the way, I've just checked and Chrome (v85) on iOS is behaving the same, with two events fired.

**Comment by johnd0e** [comment]
Ok, see what I've found:
1) Preventing `mousedown`/`pointerdown` does not cancel `click`. Only `touchstart` can be prevented.
2) But currently Leaflet has rather weird behavior: when pointer events are available - it substitute touch events with pointer events.

That's why we now have those extra clicks.

**Comment by dankarran** [comment]
If it helps as a quick fix as a stopgap (and doesn't break other things), commenting out that simulated event does seem to work for me in both Chrome and Safari on iOS...

https://github.com/Leaflet/Leaflet/blob/4f32a5e83525a3a42980c974e48712b058510eb5/src/map/handler/Map.Tap.js#L100

**Comment by bazhanius** [comment]
https://s3.amazonaws.com/leafletjs-cdn/content/leaflet/master/leaflet-src.js
^ this version seems to work way better than `1.7.1` in desktop Safari 14.0.
tap or click fires only once 👍

**Comment by lmdc45** [comment]
> 
> 
> If it helps as a quick fix as a stopgap (and doesn't break other things), commenting out that simulated event does seem to work for me in both Chrome and Safari on iOS...
> 
> https://github.com/Leaflet/Leaflet/blob/4f32a5e83525a3a42980c974e48712b058510eb5/src/map/handler/Map.Tap.js#L100

I had the 2 click event issue on all device (desktop firefox, chrome, ios safari, chrome..) and this resolved the issue

**Comment by Guttz** [comment]
The same problem happens to me and only happens in Safari. Chrome and Firefox are only firing the event once MacOS 10.15.6 and Safari 14.0.

I temporarily downgraded Leaflet to 1.6 and everything works normally.

**Comment by PeterPan73** [comment]
Is there a chance this bug will be fixed?
Does anyone work on it?

**Comment by johnd0e** [comment]
As I've already said: the bug is caused by extremely outdated `tap` handler (#6980).

To get things off the ground we need to make some decisions: https://github.com/Leaflet/Leaflet/issues/7255#issuecomment-691109158
@IvanSanchez @mourner

**Comment by IvanSanchez** [comment]
> To get things off the ground we need to make some decisions: [#7255 (comment)](https://github.com/Leaflet/Leaflet/issues/7255#issuecomment-691109158)

@johnd0e I don't have neither the time to test this, nor any iOS hardware. Anyhow, I'm trusting you on this, so I'm greenlighting #7026 (which feels cleaner than #7029)

>  Does anyone work on it?

You know what'd be cool, @PeterPan73? For you to go ahead with making a couple of builds from #7026 and #7029, and testing those for further compatibility issues, instead of just putting more pressure on the maintainers.

**Comment by johnd0e** [comment]
> so I'm greenlighting #7026 (which feels cleaner than #7029)

Unfortunately in last commit of #7026 I had to apply ugly workarounds in order to overcome [other bugs](https://github.com/Leaflet/Leaflet/pull/7026#issuecomment-606575241) (which in turn can be addressed by #7029 and #7059).

So in the end we still will have to deal with #7029 and other connected PRs, that's why I've united them all under https://github.com/Leaflet/Leaflet/projects/1

**Comment by sertal70** [comment]
I run into this issue on macOS 10.15.7 using Chrome 85 DevTools with iOS device enabled. The quickest workaround for me has been to use lodash [`debounce`](https://lodash.com/docs/4.17.15#debounce) function.

**Comment by Alisaien** [comment]
Today I ran into an issue, which seems connected to this one, so I will post my experience here.
This occurs with the Macbook trackpad and I discovered it, when trying to add a custom contextmenu to the map.
What happens is that, once the map has focus, even if you right click for the contextmenu, Leaflet will fire a simulated <click>  event, which should not be fired and has `button` value 0. This leads to weird interactions with my close handler for the contextmenu.

To produce the bug you can go to https://leafletjs.com and then:
1. Click on the map (so the map has focus)
2. "Right click"
3. Click on the map
4. "Right click"

What you will see is that the cursor is in dragging mode and once you twice click on the map (once the close contextmenu and once to trigger click event) the map will jump.

Maybe this relates also to some other issues (e.g. #6865).

**Comment by atorger** [comment]
This twice click event (on trusted on not) occurs for me on 1.7.1 when simulating mobile on Chrome and Firefox on desktop Linux (didn't happen on 1.6). Works well on real Android touch devices though (Chrome/Firefox tested). I have for now solved it by implementing my own bounce timer in the click event handler, filtering out clicks that happen within 150 ms from the last one.

**Comment by rix1** [comment]
> I have for now solved it by implementing my own bounce timer in the click event handler, filtering out clicks that happen within 150 ms from the last one.

Hi @atorger, how did you implement this? I tried `event.originalEvent.preventDefault();` in the onClick event handler, but with no success. Can you share a minimal example? 🙏 😄

**Comment by atorger** [comment]
> > I have for now solved it by implementing my own bounce timer in the click event handler, filtering out clicks that happen within 150 ms from the last one.
> 
> Hi @atorger, how did you implement this? I tried `event.originalEvent.preventDefault();` in the onClick event handler, but with no success. Can you share a minimal example? pray smile

sure:

var lastClick = Date.now();
e.addEventListener('click', function(event) {
    var now = Date.now();
    var diff = now - lastClick;
    lastClick = now;
    event.stopPropagation();
    if (diff > 150) {
        // handle the click event
    }
});

**Comment by danmichaelo** [comment]
Note that this issue also causes `bindPopup` to not work with markers:

- Version 1.7.1: Nothing happens when clicking the marker https://jsfiddle.net/danmichaelo/gnds7a8o/15/
- master: Works as expected again 🎉   https://jsfiddle.net/danmichaelo/gnds7a8o/16/

**Comment by eikes** [comment]
I ran into this problem in firefox as well (and mobile safari). I was not able to reproduce it reliably but the hint that it comes from the tap option led me to the solution which disables tap entirely as I apparently don't need it to open the popup, even on mobile safari:

```js
var map = L.map('map', { "tap": false });
```

**Comment by dartrax** [comment]
> I ran into this problem in firefox as well (and mobile safari). I was not able to reproduce it reliably but the hint that it comes from the tap option led me to the solution which disables tap entirely as I apparently don't need it to open the popup, even on mobile safari:
> 
> ```js
> var map = L.map('map', { "tap": false });
> ```

Thank you for this solution, @eikes! I ran into this issue today and that solved it.

**Comment by johnd0e** [comment]
> I apparently don't need it to open the popup, even on mobile safari:

On mobile safari `tap` handler is needed to simulate `contextmenu` event on taphold.

**Comment by tloizel** [comment]
This bug fires two touches when trying to click on map on a mobile device.

**Comment by nekromoff** [comment]
I can confirm. multiple users reported popups not working for iOS for our map using Leaflet.

**Comment by nekromoff** [comment]
See https://github.com/domoritz/leaflet-locatecontrol/issues/280 and the workaround that fixes the issue.

**Comment by jorisgeer** [comment]
This issue prevents menus from working on Safari. For example, openstreetmap loses its map tools. As no one is assigned to this bug, it may not get attention.

**Comment by binakot** [comment]
Today I faced with the same bug on the last version of the iOS. Popup appears and closes instantly due to duplicate click.

**Comment by lysdexic-audio** [comment]
disabling tap altogether fixes this for me on safari (using leaflet from react-leaflet)

```javascript
var map = L.map('map', { "tap": false });
```

works great!

source: https://bleepcoder.com/leaflet/694192486/leaflet-1-7-1-causes-2-click-events-to-be-emitted-by-leaflet

**Comment by ohjohnsen** [comment]
Disabling tap fixed the problem for me also.  But I'm using React-Leaflet, and the MapContainer component has a `tap` prop that can be used for disabling tapping.

    <MapContainer tap={false}>

**Comment by alexharris** [comment]
For anyone using Vue-Leaflet, setting tap to false seems to work to resolve the issue, but remember that vue-leaflet only allows some options to be passed as props to the map component directly, for the rest (like 'tap') you need to pass a custom 'options' object, as per: https://vue2-leaflet.netlify.app/faq/#how-can-i-specify-leaflet-options-that-aren-t-part-of-a-component-s-props

**Comment by Falke-Design** [comment]
@johnd0e @IvanSanchez is there now a fix merged or is this still open?

**Comment by johnd0e** [comment]
#7026 is meant to fix this, unmerged yet.

**Comment by justinjngan** [comment]
Hi! I've read through this thread and I see several notes about this being fixed from the discussions when the thread first started back in 2020 to more recently. 

I am still seeing this happening on Safari Desktop.  Is this expected ?

Thanks Justin

**Comment by Malvoz** [comment]
@justinjngan see https://github.com/Leaflet/Leaflet/issues/7255#issuecomment-955691965.

**Comment by alex-kowalczyk** [comment]
This bug is also reproducible on Leaflet 1.7.1 on Gnome Web (Epiphany).

**Comment by AduchiMergen** [comment]
please release 1.7.2 with this fix

**Comment by Falke-Design** [comment]
@AduchiMergen we are currently in the final phase before releasing 1.8.0. We think we release it with march

**Comment by piotr-cz** [comment]
@Falke-Design The 1.8 is quite big update which includes breaking changes
IMHO bugfix should be released in a patch update 1.7.2

**Comment by NicoHood** [comment]
If 1.8.0 includes breaking changes, why isnt it a 2.0.0 release then??

**Comment by mourner** [comment]
It has very minor breaking changes that shouldn't affect most codebases, and NPM doesn't upgrade minor versions anyway. v2.0 will have more substantial changes like dropping support for old IE.

**Comment by SpencerKaiser** [comment]
Any updates on ETA? I just spent 30 minutes debugging this only to stumble upon this issue after tracing it back to the `Marker` `onClick` 😅

**Comment by mourner** [comment]
@SpencerKaiser can you try `leaflet@beta`?

**Comment by SpencerKaiser** [comment]
@mourner one of the devs on my team (@joswayski) is still looking into it but he did find this one _breaking change_... looks like the Ukrainian flag is misaligned 🤣 Want us to open new issue for that?

Also do yall have an eta on the next release? I'm fine using the beta build for now, but I'm worried about future releases breaking us, so I'm nervous about checking that in as permanent fix.

![image](https://user-images.githubusercontent.com/6445731/161775407-aae63639-5ea8-4852-916a-0f50b9586707.png)

**Comment by mourner** [comment]
@SpencerKaiser yes, please make an issue or at least a minimal reproducible live test case! Seems like some kind of CSS conflict.

**Comment by mourner** [comment]
> Also do yall have an eta on the next release? I'm fine using the beta build for now, but I'm worried about future releases breaking us, so I'm nervous about checking that in as permanent fix.

ETA a few days. Very likely at the end of the week.

**Comment by SpencerKaiser** [comment]
The `off topic` police are going a bit overboard here 🙄 the comment directly above this one is _**directly related**_ to this issue...

**Comment by Malvoz** [comment]
👮
